### PR TITLE
Support GHC 9.0.1

### PR DESCRIPTION
**Pull request by christiaanb**
Sadly, hvr's ubuntu ghc ppa does not include the ghc-9.0.1-rc1. You can test locally with: `cabal test --allow-newer=ghc-paths:Cabal`

I don't know the stance on CPP among `hint` maintainers, so I'm willing to make any suggested style changes wrt conditional compilation.

### CODE REVIEW

**Comment by gelisam** [comment]
> I don't know the stance on `CPP` among hint maintainers, so I'm willing to make any suggested style changes wrt conditional compilation.

It's already used in several places throughout hint's codebase. Is there any other way to support more than one version of ghc?

File: src/Control/Monad/Ghc.hs
Comment (gelisam) [inline src/Control/Monad/Ghc.hs:62]: this change is pretty surprising to me! ghc has always been very conservative about which modules it depends on, but this change implies that ghc is now depending on http://hackage.haskell.org/package/exceptions?

Comment (christiaanb) [reply src/Control/Monad/Ghc.hs:62]: That's correct: https://github.com/ghc/ghc/blame/master/compiler/ghc.cabal.in#L73

Comment (gelisam) [reply src/Control/Monad/Ghc.hs:62]: huh, looks like it does!

**Comment by christiaanb** [comment]
> 
> 
> > I don't know the stance on `CPP` among hint maintainers, so I'm willing to make any suggested style changes wrt conditional compilation.
> 
> It's already used in several places throughout hint's codebase. Is there any other way to support more than one version of ghc?

Another option is to use multiple `src_XYZ` folders and use `cabal` to pick the right one:
```
if impl(ghc >= 9.0.0)
  hs-source-dirs: src_ghc9
```

**Comment by gelisam** [comment]
Thank you very much! I have way too many projects to maintain these days, so the help is very much appreciated. Let me know if you would be interested on [pairing](https://twitter.com/haskell_cat/status/1343324364968779779) on hint or Clash sometimes!

There are only two things missing, but they're easy, I can take care of them myself once I've tested and merged this. First, a CHANGELOG.md entry. Second, in order to limit the amount of work, hint's policy is currently to only support up to 3 versions of ghc, so now that we're adding 9.0.1, we should drop support for 8.4.

Oh and ideally CI should be updated to test 9.0.1, but since that version of ghc is not currently easy to obtain, I've created #110 as a reminder to do that once it does become available.

**Review by gelisam — APPROVED** [review]

**Comment by gelisam** [comment]
There doesn't seem to be any `#if`s specific to ghc 8.4, so I see no reason to drop support for 8.4 after all. I have pushed the CHANGELOG change. Do you need me to release a new version of hint on hackage?

**Comment by christiaanb** [comment]
> Thank you very much! I have way too many projects to maintain these days, so the help is very much appreciated. Let me know if you would be interested on [pairing](https://twitter.com/haskell_cat/status/1343324364968779779) on hint or Clash sometimes!

Yeah, I sorta know what you mean. With the business I don't have much time to do coding for projects myself these days, most of it is done by my other team members. I'll let you know if/when I have some time for some pair programming.

> 
> There are only two things missing, but they're easy, I can take care of them myself once I've tested and merged this. First, a CHANGELOG.md entry. Second, in order to limit the amount of work, hint's policy is currently to only support up to 3 versions of ghc, so now that we're adding 9.0.1, we should drop support for 8.4.
> 
> Oh and ideally CI should be updated to test 9.0.1, but since that version of ghc is not currently easy to obtain, I've created #110 as a reminder to do that once it does become available.

Yeah, this is an issue for some of our projects as well. Can only test locally right now.

> There doesn't seem to be any #ifs specific to ghc 8.4, so I see no reason to drop support for 8.4 after all. I have pushed the CHANGELOG change. Do you need me to release a new version of hint on hackage?

There's no rush wrt a hackage release, we can use `cabal.project` or `stack.yaml` files to manage pulling dependencies from github.

**Pull request merged**
